[Karthikeyan Vigneshram] iP

shengiv opened the pull request.

shengiv committed:
Implemented initial skeletal version of Duke

shengiv committed:
Add ability to echo user input

shengiv committed:
Add ability to store user input and display back to user

shengiv committed:
Add ability to mark and unmark tasks

jjtoh commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,47 @@
+                            + " "+ tasks.get(i).getName());
+                }
+                System.out.println();
+            }
+            else if(lineInput.startsWith("mark")){
```
Else if on the wrong line, should be on the same line as the close bracket on the line above. Spacing between keyword and open braces should also be added. (This is done multiple times)

pipipipi2002 replied:
combine it to one line.
```Java
} else if (...) {
```

Brandon-OS commented on src/main/java/Duke.java:
```diff
@@ -12,12 +12,32 @@ public static void main(String[] args) {
+
+            else if(lineInput.startsWith("unmark")){
+                int listIndex = Integer.parseInt(lineInput.substring(7));
+                Task unmarkedTask = tasks.get(listIndex-1);
```
Don't forget to add spaces in between calculations (e.g. listIndex - 1)

Brandon-OS commented on src/main/java/Duke.java:
```diff
@@ -12,12 +12,32 @@ public static void main(String[] args) {
         while(!lineInput.equals("bye")){
```
Consider adding a space after any loops or nesting (while (), if ())

Brandon-OS replied:
Consider adding a space between brackets e.g. ) and {

Brandon-OS commented on src/main/java/Duke.java:
```diff
@@ -2,7 +2,7 @@
     public static void main(String[] args) {
-        ArrayList<String> tasks = new ArrayList<String>();
+        ArrayList<Task> tasks = new ArrayList<Task>();
         String logo = "______________________________________________\n";
```
Consider using capitalized letters for constants (LOGO)

Brandon-OS reviewed (commented):
Overall, awesome job in making the code efficient and readable

pipipipi2002 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,47 @@
+                System.out.println("OK, I've marked this task as not done yet: ");
+                System.out.println("  " + unmarkedTask.getCheckBox() + " " + unmarkedTask.getName());
+                System.out.println();
+            }
+
+            else{
```
Space before the `{`

pipipipi2002 replied:
Combine to one line

pipipipi2002 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,47 @@
+                System.out.println("Nice! I've marked this task as done: ");
+                System.out.println("  " + markedTask.getCheckBox() + " " + markedTask.getName());
+                System.out.println();
+            }
+
+            else if(lineInput.startsWith("unmark")){
```
combine into one line

pipipipi2002 replied:
Add space before `(...)` and between `))` and `{`

pipipipi2002 reviewed (commented):
Good Job!

jjtoh commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,28 @@
+public class Task {
```
Methods are well named as they follow camelCase

jjtoh commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,28 @@
+    private String name;
+    private boolean isCompleted;
+
+    public Task(String name){
```
There should be spacing before braces

pipipipi2002 replied:
add space between the ) and {

jjtoh commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,28 @@
+        this.isCompleted = false;
+    }
+
+    public String getName(){
```
Good naming for setters and getters

pipipipi2002 replied:
add space between ) and {

jjtoh reviewed (commented):
Just some repeated spacing mistakes, otherwise it is generally well done.

pipipipi2002 commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,28 @@
+        return name;
+    }
+
+    public boolean getIsCompleted(){
```
add space between ) and {

pipipipi2002 commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,28 @@
+        return isCompleted;
+    }
+
+    public void setIsCompleted(boolean isCompleted){
```
add space between ) and {

pipipipi2002 commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,28 @@
+        this.isCompleted = isCompleted;
+    }
+
+    public String getCheckBox(){
```
add space between ) and {

pipipipi2002 commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    public String getCheckBox(){
+        if(isCompleted){
```
add space between ) and {

pipipipi2002 reviewed (commented):
the spaces are most of the errors found

shengiv committed:
Tweak to comply with Java coding standards

shengiv committed:
Split Tasks into Todos, Deadlines and Events

shengiv committed:
Tweak code to improve code quality

The different tasks cause code duplication when adding the tasks to the list
Extracting the print statements to one method reduces redundancy and code duplication

Let's pull up common behaviours to methods so that the methods can be reused across different tasks.

The main method is long and difficult to read.
Due to the many conditionals present to handle different tasks, the code can get quite messy.

Let's extract the statements within the conditional statements into more readable methods.

xseh commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,144 @@
+    private static ArrayList<Task> tasks = new ArrayList<Task>();
+    private static final String LINE_BREAK = "______________________________________________\n";
+
+    public static void printList() { //prints list of tasks
```
Please consider removing the comment to avoid repeating the obvious.

xseh commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,144 @@
+    public static void printList() { //prints list of tasks
+        System.out.println(LINE_BREAK);
+        for (int i = 0; i < tasks.size(); i++) {
+            System.out.println( Integer.toString(i + 1) + "." + tasks.get(i).toString());
```
Avoid complicated expressions. Perhaps you can consider abstrating the logic into multiple statements and methods to make the code more readable.

xseh commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,144 @@
+    }
+
+    public static ToDo getToDo(String[] inputWords) { //returns a todo object from the user input sentence
+        String[] taskNameWords = Arrays.copyOfRange(inputWords, 1, inputWords.length);
```
Avoid using magic numbers. Consider putting the numbers into a descriptive variable. For example, `1` can be put into the variable `TASK_NAME_INDEX` or other appropriate names.

xseh commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,144 @@
+    }
+
+    public static Deadline getDeadline(String[] inputWords) { //returns a Deadline object from the user input sentence
+        int indexOfBy = Arrays.asList(inputWords).indexOf("/by");
```
Avoid magic strings. Consider putting the strings into a descriptive variable. For example, `/by` can be put into the variable `DEADLINE_DATA_PREFIX_BY`.

xseh commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,144 @@
+    }
+
+    public static void addDeadline(String[] inputWords) { //adds deadline to list of tasks if keyword "/by" is found
+        if(!Arrays.asList(inputWords).contains("/by")) {
```
Consider abstracting similiar logic (e.g. checking for `/by` and `/at`) into a method to avoid repeating the code and make the code easier to understand.

xseh commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,144 @@
+    }
+
+    public static void addEvent(String[] inputWords) { //adds event to list of tasks if keyword "/at" is found
+        if(!Arrays.asList(inputWords).contains("/at")) {
```
This code violates coding standards, please consider adding spaces between `if` and `(`. 

As stated in the module website, the if-else class of statements should have the following form:

> if (condition) {
    statements;
}

xseh reviewed (commented):
Good attempt! Just please consider more abstraction (SLAP harder), coding standards (particularly the spaces), and removing uneeded comments that repeats the obvious. Also take note of magic literals.

shengiv committed:
Add exceptions to handle errors

Duke accepts null descriptions.

We do not want to accept null descriptions for the tasks since each task should at least have a task name.

Creating a new exception class and throwing the exception can prevent null entries in the tasks list.

Let's create a new exception NoDescriptionException to handle such exceptions.

shengiv committed:
Add Invalid commands

Commands that are not recognized by Duke get added as tasks in the list which is not desirable.

Invalid commands should be rejected.

Let's add a print statement to notify user of the invalid command instead.

shengiv committed:
Merge branch 'branch-Level-5'

shengiv committed:
Improve code quality

shengiv committed:
Add support for deleting tasks

shengiv committed:
Add ability to save changes in task list to hard disk.

shengiv committed:
Merge branch 'branch-Level-6'

shengiv committed:
Merge branch 'branch-Level-7'

shengiv committed:
Add duke package

shengiv committed:
Extract Ui, Storage, Parser and TaskList classes from Duke class

shengiv committed:
Add ability to find tasks using keywords

shengiv committed:
Merge pull request from shengiv/branch-Level-9

Add ability to find tasks using keywords

shengiv committed:
Add JavaDoc comments

shengiv committed:
Merge pull request from shengiv/branch-A-JavaDoc

Merge branch-A-JavaDoc with master branch

shengiv committed:
Update User Guide
